Source/DafnyRuntime/DafnyRuntime.cs

```diff
@@ -1552,11 +1553,69 @@ public struct BigRational {
       num = new BigInteger(n);
       den = BigInteger.One;
     }
+    public BigRational(uint n) {
+      num = new BigInteger(n);
+      den = BigInteger.One;
+    }
+    public BigRational(long n) {
+      num = new BigInteger(n);
+      den = BigInteger.One;
+    }
+    public BigRational(ulong n) {
+      num = new BigInteger(n);
+      den = BigInteger.One;
+    }
     public BigRational(BigInteger n, BigInteger d) {
       // requires 1 <= d
       num = n;
       den = d;
     }
+    /// <summary>
+    /// Construct an exact rational representation of a double value.
+    /// Throw an exception on NaN or infinite values. Does not support
+    /// subnormal values, though it would be possible to extend it to.
+    /// </summary>
+    public BigRational(double n) {
+      if (Double.IsNaN(n)) {
+        throw new ArgumentException("Can't convert NaN to a rational.");
+      }
+      if (Double.IsInfinity(n)) {
+        throw new ArgumentException(
+          "Can't convert +/- infinity to a rational.");
+      }
+      if (Double.IsSubnormal(n)) {
+        throw new ArgumentException(
+          "Can't convert a subnormal value to a rational (yet).");
+      }
+
+      // Double-specific values
+      const int exptBias = 1023;
```

Comment:
Can you comment on the choice of algorithm?  The one implementation I'm familiar with is https://github.com/python/cpython/blob/da6c78584b1f45ce3766bf7f27fb033169715292/Objects/floatobject.c#L1560

Reply:
I derived it mostly from first principles based on the IEEE standard. I did look [here](https://github.com/ghc/ghc/blob/master/libraries/base/GHC/Float.hs) to some extent, but didn't follow it entirely.

Reply:
It'd honestly likely be faster if I followed that Haskell code more precisely, because it has no iteration and no multiplication.